File: frontend/src/utils/frequencyUtils.js
export const DAYS_PER_FREQUENCY = {
  daily: 1,
  weekly: 7,
  biweekly: 14,
  '4-week': 28,
  monthly: 30,
  quarterly: 90,
  yearly: 365,
};

export const FREQUENCY_CODES = {
  daily: 'D',
  weekly: 'W',
  biweekly: 'B',
  '4-week': '4',
  monthly: 'M',
  quarterly: 'Q',
  yearly: 'Y',
};

export function convertAmount(amount, fromFreq, toFreq) {
  if (fromFreq === toFreq) {
    return parseFloat(amount);
  }
  
  const amountNum = parseFloat(amount);
  const fromDays = DAYS_PER_FREQUENCY[fromFreq];
  const toDays = DAYS_PER_FREQUENCY[toFreq];
  
  const dailyRate = amountNum / fromDays;
  const result = dailyRate * toDays;
  
  return result;
}

export function formatAmountDisplay(amount) {
  const num = parseFloat(amount);
  const rounded = Math.round(num * 100) / 100;
  
  if (Number.isInteger(rounded)) {
    return rounded.toString();
  }
  
  return rounded.toFixed(2).replace(/\.?0+$/, '');
}

export function calculatePocketInFrequency(pocketAmount, pocketFreq, items, targetFreq) {
  let total = convertAmount(pocketAmount, pocketFreq, targetFreq);
  
  items.forEach(item => {
    if (!item.is_other) {
      const itemConverted = convertAmount(item.amount, item.frequency, targetFreq);
      total += itemConverted;
    }
  });
  
  return total;
}

export function hasMixedFrequencies(pocketFreq, items) {
  if (pocketFreq === 'monthly' && items.every(item => item.frequency === 'monthly' || item.is_other)) {
    return false;
  }
  
  return items.some(item => !item.is_other && item.frequency !== pocketFreq);
}

File: backend/api/frequency_utils.py
from decimal import Decimal

DAYS_PER_FREQUENCY = {
    'daily': Decimal('1'),
    'weekly': Decimal('7'),
    'biweekly': Decimal('14'),
    '4-week': Decimal('28'),
    'monthly': Decimal('30'),
    'yearly': Decimal('365'),
}

FREQUENCY_CODES = {
    'daily': 'D',
    'weekly': 'W',
    'biweekly': 'B',
    '4-week': '4',
    'monthly': 'M',
    'yearly': 'Y',
}


def convert_amount(amount, from_freq, to_freq):
    """Convert amount from one frequency to another"""
    if from_freq == to_freq:
        return Decimal(str(amount))
    
    amount_decimal = Decimal(str(amount))
    from_days = DAYS_PER_FREQUENCY[from_freq]
    to_days = DAYS_PER_FREQUENCY[to_freq]
    
    daily_rate = amount_decimal / from_days
    result = daily_rate * to_days
    
    return result


def format_amount_display(amount):
    """Format amount to 2 decimals for display"""
    amount_decimal = Decimal(str(amount))
    rounded = amount_decimal.quantize(Decimal('0.01'))
    return str(rounded)


def calculate_pocket_monthly_equivalent(pocket_amount, pocket_freq, items):
    """Calculate monthly equivalent of pocket including all items"""
    monthly_total = convert_amount(pocket_amount, pocket_freq, 'monthly')
    
    for item in items:
        if not item.is_other:
            item_monthly = convert_amount(item.amount, item.frequency, 'monthly')
            monthly_total += item_monthly
    
    return monthly_total


def calculate_pocket_in_frequency(pocket_amount, pocket_freq, items, target_freq):
    """Calculate pocket amount in target frequency including all items"""
    monthly_total = calculate_pocket_monthly_equivalent(pocket_amount, pocket_freq, items)
    return convert_amount(monthly_total, 'monthly', target_freq)


def has_mixed_frequencies(pocket_freq, items):
    """Check if pocket and items have different frequencies"""
    if pocket_freq == 'monthly' and all(item.frequency == 'monthly' or item.is_other for item in items):
        return False
    
    for item in items:
        if not item.is_other and item.frequency != pocket_freq:
            return True
    
    return False